ruma-common/src/push.rs

```diff
@@ -80,6 +85,97 @@ impl Ruleset {
             AnyPushRule::Sender(r) => self.sender.insert(r),
         }
     }
+
+    /// Get the push actions that apply to this event.
+    ///
+    /// Returns `None` if no push action applies.
+    ///
+    /// # Arguments
+    ///
+    /// * `event` - The raw JSON of a room message event.
+    /// * `context` - The context of the message and room at the time of the event.
+    pub fn get_actions<'a, T>(
+        &'a self,
+        event: &Raw<T>,
+        context: &PushConditionRoomCtx,
+    ) -> Option<impl Iterator<Item = &'a Action>> {
```

Comment:
I wonder whether this should return just `impl Iterator<Item = &'a Action>>` and simply yield no items if no condition applies.

cc @timokoesters

Reply:
I did that because I didn't know how to return an empty Iterator.

Reply:
https://doc.rust-lang.org/std/iter/fn.empty.html

Reply:
I think you can just add `.flatten()` to every return `return Some(_)`. And `None.flatten()` as the final return value to match up the types. Otherwise:

* Define a type, e.g. `ActionIter` (can / should be private since we "hide" it behind `impl Iterator`) as an enum over `{ Empty, List(vec::Iter<'a>) }` (or similar)
* `impl<'a> Iterator for ActionIter<'a>` with `next()` returning `None` for `Empty`, `list.next()` for `List(list)`
* Return `ActionIter::List(rule.actions)` instead of `Some(rule.actions)`

Reply:
@DevinR528 That would help if we were boxing the iterator, but there's not really a need to do that here.

Reply:
It didn't work with `flatten`. However I noticed the `actions.iter()` were returning a slice so I put `[].iter()` as final value and it works

Reply:
I meant `Some(actions.iter()).flatten()`, not `actions.iter().flatten()`. But `[].iter()` seems better 👍🏼

Reply:
I know, and I actually tried everything: in, out, without iter… When I put it like you said I get:

```
the method `flatten` exists for enum `std::option::Option<std::slice::Iter<'_, Action>>`, but its trait bounds were not satisfied
the following trait bounds were not satisfied:
`std::option::Option<std::slice::Iter<'_, Action>>: Iterator`
which is required by `&mut std::option::Option<std::slice::Iter<'_, Action>>: Iterator`
```

Reply:
Oh, right. Would have been `Some(actions.iter()).into_iter().flatten()`.